### ISSUE TITLE
Linux (DEB) Package Creation

### ISSUE THREAD

**Issue by vikman90** [wazuh/wazuh-agent#143]
**Parent Issue**:
- https://github.com/wazuh/wazuh-agent/issues/128

## Description
This issue focuses on creating a DEB package for Debian-based distributions such as Debian and Ubuntu. The package should manage all necessary dependencies and include pre/post-installation scripts to ensure that the Wazuh agent service is automatically enabled and running after installation.

## Functional Requirements
- Create a DEB package compatible with Debian-based distributions like Ubuntu and Debian.
- Ensure all dependencies are correctly handled during the installation process.
- - The Wazuh agent service must be disabled by default.
- Test the installation and uninstallation workflows to ensure compatibility with Debian-based systems.

## Acceptance Criteria
- The DEB package is built successfully and integrates with native package managers.
- Installation, uninstallation, and upgrade processes are tested and verified on Debian-based distributions.

**Comment by lchico** [comment]
# Update
I've begun researching this issue and have started working on a solution.

**Comment by lchico** [comment]
# Update
- Researched how to generate packages on Debian and identified best practices.
I have detected some inconsistencies in my Dockerfile that are preventing the successful build of GCC 14. I am comparing it with a working container to find the differences.
- I created a [slack thread](https://wazuh-team.slack.com/archives/C06HWMDP476/p1726798627793369) outlining the implementation.
- Encountered a dependency issue where the first round of package generation works on Ubuntu but fails on Debian:

```
/opt/wazuh-agent/bin/wazuh-agent: /lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.32' not found (required by /opt/wazuh-agent/bin/wazuh-agent)
/opt/wazuh-agent/bin/wazuh-agent: /lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.31' not found (required by /opt/wazuh-agent/bin/wazuh-agent)
/opt/wazuh-agent/bin/wazuh-agent: /lib/x86_64-linux-gnu/libstdc++.so.6: version `CXXABI_1.3.15' not found (required by /opt/wazuh-agent/bin/wazuh-agent)
```
I am not prioritizing this issue for now as there are other things that need to be defined first.

**Comment by vikman90** [comment]
Dear @lchico, let me write here a brief of our discussion:

1. **Service Definition:**
   The [`man 7 file-hierarchy`](https://man7.org/linux/man-pages/man7/file-hierarchy.7.html) document clearly indicates that the `/usr` directory is reserved for vendor-provided packages. Therefore, it's more appropriate for us to maintain the service file under `/etc/systemd/system/`.

2. **User and Group:**
   We haven't defined a change in user/group yet, but it's likely we will do so via Systemd in the future. For now, we should continue using `wazuh` for both the user and group. This may change later, but we can address it at that point.

3. **Configuration Files:**
   Following the vendor recommendations is the best approach here. I propose we use `/etc/wazuh-agent/wazuh-agent.yml` for the main configuration file. This aligns with the structure of other files (like those for OpenSearch), and allows us to manage multiple files if needed. Additionally, this would allow for both a manager and an agent to coexist on the same system.

4. **Generated Files Location:**
   For generated files, it's more appropriate to place them in directories like `/var/local/wazuh-agent` or `/var/opt/wazuh-agent`, rather than `/opt`, which is reserved for static files. We'll need to follow system recommendations here as well.

5. **Logs:**
   Currently, agent logs will be sent to `stdout` and captured by Systemd via Journald. In the future, we may define an option to write logs to files, in which case using a directory like `/var/log/wazuh-agent/*.log` might be more suitable. This isn't yet fully defined, though.

6. **Dependencies:**
   This requires further discussion. Since we're moving to a single binary, it would be best to statically link as many libraries as possible. If certain libraries are absolutely necessary, we may need to ship them with the agent (similar to how we handle `/var/ossec/lib`). The most problematic libraries here may be `libstdc++` and `libgcc_s`.

7. **Removing Package:**
   I agree with your approach: remove `/etc/opt/wazuh-agent` but keep `/etc/wazuh.conf`. For backups, there's nothing crucial that we need to preserve for now. However, it may make sense to remove the `wazuh` user as well, since it's only required at runtime.

8. **Purge:**
   The purge process looks good. Removing `/etc/opt/wazuh-agent`, `/etc/wazuh.conf`, and the `wazuh` user/group should suffice.

---

In conclusion, for all of these items, we need to ensure we follow the system standards for each platform (Debian, RedHat, Windows, macOS), reference the vendor's documentation, and review our requirements. We should adhere to the default standards unless we have a clear reason to deviate, which should be documented thoroughly.

### References

- [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs-3.0.html)
- [systemd.service — Service unit configuration](https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html)

**Comment by lchico** [comment]
First, thank you @vikman90. That sounds very clear to me.

# Update
- Fixed the Docker issue, and I can now build a deb package.
- Updated Debian scripts and successfully generated the first Debian package with the binary.

**Comment by mjcr99** [comment]
## Update
**(24/09/2024)** Updating Dockerfile to enable compiling with the new repository, covering different dependencies. The GCC version has been downgraded to 13.2.0 since using 14.2 was causing `procps` port-overlay to not compile.

**Comment by MarcelKemp** [comment]
## Update
Test branch: `test/deb-package-generation`

In the branch, I have rebased the rest of the branches and added some steps to the Dockerfile to try to generate a 5.0 package.
It still needs modifications to add the service functionality.
